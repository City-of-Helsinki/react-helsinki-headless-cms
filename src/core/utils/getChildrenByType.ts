import React from 'react';

/**
 * Gets the string type of the component or core html (JSX) element.
 * React Fragments will return type 'react.fragment'.
 * Priority will be given to the prop '__TYPE'.
 */
export const typeOfComponent = (component: React.ReactElement): string =>
  // eslint-disable-next-line no-underscore-dangle
  component?.props?.__TYPE ||
  component?.type
    ?.toString()
    .replace('Symbol(react.fragment)', 'react.fragment') ||
  undefined;

/**
 * Gets all children by specified type.
 * This function will check the prop '__TYPE' first
 * and then the 'type' string to match core html elements.
 * To find a React Fragment, search for type 'react.fragment'.
 *
 * @example
 * // Finds all occurrences of ToDo (custom component), div, and React Fragment
 * getChildrenByType(children, ['ToDo', 'div', 'react.fragment']);
 */
export const getChildrenByType = (
  children: React.ReactNode,
  types: string[],
): React.ReactNode[] =>
  React.Children.toArray(children).filter(
    (child) =>
      React.isValidElement(child) &&
      types.indexOf(typeOfComponent(child)) !== -1,
  );
