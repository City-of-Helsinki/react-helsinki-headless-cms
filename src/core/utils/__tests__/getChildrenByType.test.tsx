import React from 'react';

import { getChildrenByType } from '../getChildrenByType';

describe('getChildrenByType', () => {
  it.each([
    [
      ['img', 'p'],
      <>
        <p>should contain 1/3</p>
        <img alt="should contain 2/3" src="#" />
        <span>should not contain</span>
        <p>should contain 3/3</p>
      </>,
      3,
    ],
    [
      ['ul'],
      <>
        <ol>
          <li>Should not contain</li>
        </ol>
        <ul>
          <li>Should contain 1/2</li>
        </ul>
        <p>Should not contain</p>
        <hr />
        <ul>
          <li>Should contain 2/2</li>
        </ul>
      </>,
      2,
    ],
  ])(
    'filters childrens by defined list of types',
    (types, component, filteredCount) => {
      const { children } = component.props;
      const result = getChildrenByType(children, types);
      expect(result).toHaveLength(filteredCount);
      expect(result).toMatchSnapshot();
    },
  );
});
