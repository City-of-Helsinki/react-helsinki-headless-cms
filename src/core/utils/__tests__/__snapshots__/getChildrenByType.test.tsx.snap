// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getChildrenByType filters childrens by defined list of types 1`] = `
[
  <p>
    should contain 1/3
  </p>,
  <img
    alt="should contain 2/3"
    src="#"
  />,
  <p>
    should contain 3/3
  </p>,
]
`;

exports[`getChildrenByType filters childrens by defined list of types 2`] = `
[
  <ul>
    <li>
      Should contain 1/2
    </li>
  </ul>,
  <ul>
    <li>
      Should contain 2/2
    </li>
  </ul>,
]
`;
